hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHADBTransactionBufferImpl.java

```diff
@@ -92,43 +104,54 @@ public void setLatestSnapshot(SnapshotInfo latestSnapshot) {
 
   @Override
   public void flush() throws IOException {
-    // write latest trx info into trx table in the same batch
-    Table<String, TransactionInfo> transactionInfoTable
-        = metadataStore.getTransactionInfoTable();
-    transactionInfoTable.putWithBatch(currentBatchOperation,
-        TRANSACTION_INFO_KEY, latestTrxInfo);
-
-    metadataStore.getStore().commitBatchOperation(currentBatchOperation);
-    currentBatchOperation.close();
-    this.latestSnapshot = latestTrxInfo.toSnapshotInfo();
-    // reset batch operation
-    currentBatchOperation = metadataStore.getStore().initBatchOperation();
-
-    DeletedBlockLog deletedBlockLog = scm.getScmBlockManager()
-        .getDeletedBlockLog();
-    Preconditions.checkArgument(
-        deletedBlockLog instanceof DeletedBlockLogImpl);
-    ((DeletedBlockLogImpl) deletedBlockLog).onFlush();
+    rwLock.writeLock().lock();
+    try {
+      // write latest trx info into trx table in the same batch
+      Table<String, TransactionInfo> transactionInfoTable
+          = metadataStore.getTransactionInfoTable();
+      transactionInfoTable.putWithBatch(currentBatchOperation,
+          TRANSACTION_INFO_KEY, latestTrxInfo);
+
+      metadataStore.getStore().commitBatchOperation(currentBatchOperation);
+      currentBatchOperation.close();
+      this.latestSnapshot = latestTrxInfo.toSnapshotInfo();
+      // reset batch operation
+      currentBatchOperation = metadataStore.getStore().initBatchOperation();
+
+      DeletedBlockLog deletedBlockLog = scm.getScmBlockManager()
+          .getDeletedBlockLog();
+      Preconditions.checkArgument(
+          deletedBlockLog instanceof DeletedBlockLogImpl);
+      ((DeletedBlockLogImpl) deletedBlockLog).onFlush();
+    } finally {
+      rwLock.writeLock().unlock();
+    }
   }
 
   @Override
   public void init() throws IOException {
     metadataStore = scm.getScmMetadataStore();
```

Comment:
Why is this assignment outside of the protected block?

Reply:
This assignment is idempotent, so I just left it alone. It is OK to move inside. @adoroszlai

Reply:
Thanks @Xushaohong for the explanation, it's OK to leave it as is.